Objects/rangeobject.c

```diff
@@ -171,6 +171,41 @@ range_dealloc(rangeobject *r)
     PyObject_Free(r);
 }
 
+static unsigned long
+get_len_of_range(long lo, long hi, long step);
+
+/* Return the length as a long, -1 for an overflow and -2 for any other type of error
+ *
+ * Exceptions are cleared
+ */
+static long compute_range_length_long(PyObject *start,
+                PyObject *stop, PyObject *step) {
+    int overflow = 0;
+
+    long long_start = PyLong_AsLongAndOverflow(start, &overflow);
+    if (overflow)
+        return -1;
+    if (long_start==-1 && PyErr_Occurred()) {
+        PyErr_Clear();
```

Comment:
With apologies for fussing about a case that currently can't even happen: this still isn't quite right. If we get an _unexpected_ exception from `PyLong_AsLongAndOverflow` (and right now, _any_ exception from `PyLong_AsLongAndOverflow` would count as unexpected), then we'll want to propagate that to the caller rather than clearing it. Otherwise we're doing the C-API equivalent of a Python "except: pass".

So I _think_ all that's needed is to drop the `PyErr_Clear()` here, and check for a return of `-2` in the calling function.

To avoid too much confusion, we could also consider swapping the return values around (so `-2` means overflow and `-1` means unexpected exception), since returning `-1` for an exception is fairly consistent in the rest of the codebase.

Reply:
You have a point. My line of reasoning was to bail out in case of any error, clear the exception and then continue with the regular path (which could encounter and then handle the same error). I updated the PR to not clear the error, but check for the return value in the calling function. In the hypothetical case (since we both agree it cannot happen) such an error is propagated, we can either clear the error in the caller or propagate again. I have chosen the latter.

The choice for -1 was because that is used for overflow in `PyLong_AsLongAndOverflow`. But `PyLong_AsLongAndOverflow` also uses -1 for error checking, so agree that -1 is more in line with the rest of the code.

I also added assert statements to check that in `compute_range_length` the arguments have `PyLong_Check` equal to 1.

Reply:
@mdickinson The reasoning about the case that cannot happen is actually quite useful, as there was a bug in the code. The problem was that `get_len_of_range` guarantees the results fits into an unsigned long, but we cast to long without any checks. I added a check and a regression test. (not sure what the correct way is to check this is cpython, I just checked the result of the cast is negative)

Reply:
> (not sure what the correct way is to check this is cpython, I just checked the result of the cast is negative)

Thanks. That's not totally safe in standard C, since conversions from unsigned to signed are implementation-defined. (C99 §6.3.1.3p3). I think we want to emulate what the other callers of `get_len_of_range` do and compare the return value to `LONG_MAX`.

I took the liberty of pushing a commit that does this, along with a couple of drive-by style and consistency fixes. If you're okay with the latest commit, I'll merge this PR once CI completes. And if not, I'm happy to revert and/or discuss further.

Reply:
@mdickinson Thanks for the improvement with the cast. Changes are fine with me.